OpenRA.Mods.Common/Widgets/SupportPowerTimerWidget.cs

```diff
@@ -11,55 +11,150 @@
 using System.Collections.Generic;
 using System.Drawing;
 using System.Linq;
+using OpenRA.Graphics;
 using OpenRA.Mods.Common.Traits;
-using OpenRA.Primitives;
 using OpenRA.Widgets;
 
 namespace OpenRA.Mods.Common.Widgets
 {
 	public class SupportPowerTimerWidget : Widget
 	{
-		public readonly string Font = "Bold";
-		public readonly string Format = "{0}: {1}";
-		public readonly TimerOrder Order = TimerOrder.Descending;
+		[Translate] public readonly string NoTeamText = "No Team";
+		[Translate] public readonly string TeamText = "Team {0}";
 
-		readonly IEnumerable<SupportPowerInstance> powers;
-		Pair<string, Color>[] texts;
+		readonly Dictionary<Player, SupportPowerManager> spManagers;
+		readonly SortedDictionary<int, Player[]> playersByTeam;
 
-		[ObjectCreator.UseCtor]
-		public SupportPowerTimerWidget(World world)
+		readonly string fontName = "Bold";
+		readonly SpriteFont font;
+		readonly float yIncrement, powerPos, timePos;
+
+		// Represents one line of text
+		struct Text
 		{
-			powers = world.ActorsWithTrait<SupportPowerManager>()
-				.Where(p => !p.Actor.IsDead && !p.Actor.Owner.NonCombatant)
-				.SelectMany(s => s.Trait.Powers.Values)
-				.Where(p => p.Instances.Any() && p.Info.DisplayTimer && !p.Disabled);
+			public Color PlayerColor;
+			public string Player;
+			public string Power;
+			public string RemainingTime;
+			public bool Ready;
```

Comment:
Instead of having a separate Ready variable here, I could compare RemainingTime to zero time. I don't know which one is better.